shared-bindings/bitmaptools/__init__.c

```diff
@@ -296,6 +296,68 @@ STATIC mp_obj_t bitmaptools_obj_fill_region(size_t n_args, const mp_obj_t *pos_a
 }
 
 MP_DEFINE_CONST_FUN_OBJ_KW(bitmaptools_fill_region_obj, 0, bitmaptools_obj_fill_region);
+//|
+//| def boundary_fill(
+//|        dest_bitmap: displayio.Bitmap,
+//|        x: int, y: int,
+//|        value: int, background_value: int) -> None:
+//|      """Draws the color value into the destination bitmap enclosed
+//|      area of pixels of the background_value color. Like "Paint Bucket"
+//|      fill tool.
+//|
+//|      :param bitmap dest_bitmap: Destination bitmap that will be written into
+//|      :param int x: x-pixel position of the first pixel to check and fill if needed
+//|      :param int y: y-pixel position of the first pixel to check and fill if needed
+//|      :param int value: Bitmap palette index that will be written into the
+//|             enclosed area in the destination bitmap
+//|      :param int background_value: Bitmap palette index that will filled with the
+//|             value color in the enclosed area in the destination bitmap"""
+//|      ...
+//|
+STATIC mp_obj_t bitmaptools_obj_boundary_fill(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum {ARG_dest_bitmap, ARG_x, ARG_y, ARG_value, ARG_background_value};
+
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_dest_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_x, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_y, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_value, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_background_value, MP_ARG_REQUIRED | MP_ARG_INT},
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_bitmap_t *destination = MP_OBJ_TO_PTR(mp_arg_validate_type(args[ARG_dest_bitmap].u_obj, &displayio_bitmap_type, MP_QSTR_dest_bitmap)); // the destination bitmap
+
+    uint32_t value, color_depth;
+    value = args[ARG_value].u_int;
+    color_depth = (1 << destination->bits_per_value);
+    if (color_depth <= value) {
+        mp_raise_ValueError(translate("value out of range of target"));
+    }
+
+    uint32_t background_value;
+    background_value = args[ARG_background_value].u_int;
+    if (color_depth <= background_value) {
+        mp_raise_ValueError(translate("background value out of range of target"));
+    }
+
```

Comment:
```suggestion
    if (value == background_value) {
        // There is nothing to do
        return mp_const_none;
    }
```

If the user passes in the same value for `value` and `background_value` you will get an infinite loop and a MemoryError. This extra check prevents that. Tested on a PyPortal.

Reply:
I agree about the naming of the parameter. What do you think about more specific ones `fill_color_value` and `replaced_color_value`

I like the idea of targeting the color at the location there, but I'd to provide the protection of also passing in a target color (whatever it gets called) if the user wants it. I'll try to make it so that parameter is optional, and if not passed it will target whatever color is at the location.

Good catch on the infinite loop. Thank you!